vmdb/app/models/miq_alert.rb

```diff
@@ -150,12 +150,9 @@ def self.evaluate_hourly_timer
       prof.miq_alerts.each do |a|
         next unless a.enabled? && a.responds_to_events && a.responds_to_events.include?(HOURLY_TIMER_EVENT)
 
-        model = a.db.constantize
-        unless model.name == "Storage" # No clean way to get zone from storage
-          targets += model.includes(:ext_management_system).select { |i| i.my_zone == zone }
-        else
-          targets += model.all
-        end
+        # Zone has accessors to these related resources:
+        # ems_clusters, storages, hosts, ext_management_systems, miq_servers, vms
+        targets += zone.public_send(a.db.constantize.table_name, true)
```

Comment:
Passing true here is problematic.  For the methods that you updated it means include disconnected resource.  For a few resources (miq_servers and ext_management_system) you are calling an active record relationship and unintentionally telling it to discard any cache copy of that association it might have.  This feels wrong.

What I am thinking we could do instead is move the new logic you created for determining disconnected resource into new class methods, like vms_without_a_zone and hosts_with_a_zone.  The existing zone methods would return to their original states and we could do the aggregation of the two collections here.  The only additional thing we would need here is a check that the zone responds to the "#{table_name}_with_a_zone" method to determine if we were going to make the call for disconnected resources for that type.

Then if we do eventually break out the disconnected resources into their own scheduled task we would already have the isolated methods to determine the resources and removing them from here would involve deleting a few lines.

@bzwei @Fryguy What do you think?